test/jdk/java/lang/StableValue/JEP.md

```diff
@@ -0,0 +1,561 @@
+
+## Goals
+
+- Provide an easy and intuitive API to describe value holders that can change at most once.
+- Decouple declaration from initialization without significant footprint or performance penalties.
+- Reduce the amount of static initializer and/or field initialization code.
+- Uphold integrity and consistency, even in a multithreaded environment.
+
+## Non-goals
+
+- It is not a goal to provide additional language support for expressing lazy computation.
+This might be the subject of a future JEP.
+- It is not a goal to prevent or deprecate existing idioms for expressing lazy initialization.
+
+## Motivation
+
+Java allows developers to control whether fields are mutable or not. Mutable fields can be updated multiple times, and
+from any arbitrary position in the code. Conversely, immutable fields (i.e. `final` fields), can only be updated
+_once_, and only in very specific places: the class initializer (for a static immutable field) or the class constructor
+(for an instance immutable field). Unfortunately, in Java there is no way to define a field that can be updated _at most
+once_ (i.e. fields that are either not updated at all or are updated exactly once) and from _any_ arbitrary position in
+the code.
+
+| Field kind         | #Updates | Code update location              |
+|--------------------|----------|-----------------------------------|
+| Mutable            | [0, ∞)   | Anywhere                          |
+| `final`            | 1        | Constructor or static initializer |
+| at-most-once (N/A) | [0, 1]   | Anywhere                          |
+
+"At-most-once fields" would be essential to expensive cache computations associated with method calls, so that they can
+be reused across multiple calls. For instance, creating a logger or reading application configurations from an external
+database. Furthermore, if the VM is made aware, a field is an "at-most-once field" and it is set, it may
+[constant-fold](https://en.wikipedia.org/wiki/Constant_folding) the field value, thereby providing crucial performance
+and energy efficiency gains. 
+
+It is also important to stress a method called to compute a value might have intended or
+unintended side effects and therefore, it would be vital to also guarantee the method is invoked at most once, even in
+a multithreaded environment.
+
+Another property of at-most-once fields would be, they are written to at most once but
+are likely read at many occasions. Hence, updating the field is not so time-critical whereas every effort to make 
+reading the field performant should be made.  
+
+Here is how a naïve cache could look like using a mutable field and where a `Logger` instance is cached:
+
+```
+// A naïve cache. Do not use this solution!
+public class Cache {
+
+    private Logger logger;
+
+    public Logger logger() {
+        Logger v = logger;
+        if (v == null) {
+            logger = v = Logger.getLogger("com.company.Foo");
+        }
+        return v;
```

Comment:
```suggestion
        if (logger == null) {
            logger = Logger.getLogger("com.company.Foo");
        }
        return logger;
```
This is the more common naive pattern. Lots of Java programmers have not seen the trick of reading a field to local variable for thread safety, not to mention the chained `=` assignments.